smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/extension/MediatorManager.java

```diff
@@ -156,68 +156,87 @@ public void initializeAndRun() {
         if (initialized) {
             throw ex.illegalStateForMediatorManagerAlreadyInitialized();
         }
+
         log.deploymentDoneStartProcessing();
 
-        streamRegistars.stream().forEach(ChannelRegistar::initialize);
+        List<ChannelRegistar> registars = streamRegistars.stream().collect(Collectors.toList());
+        for (ChannelRegistar registar : registars) {
+            registar.initialize();
+        }
         Set<String> unmanagedSubscribers = channelRegistry.getOutgoingNames();
         log.initializingMediators();
-        collected.mediators()
-                .forEach(configuration -> {
-
-                    AbstractMediator mediator = createMediator(configuration);
-
-                    log.initializingMethod(mediator.getMethodAsString());
-
-                    mediator.setDecorators(decorators);
-                    mediator.setHealth(health);
-                    mediator.setWorkerPoolRegistry(workerPoolRegistry);
-
-                    try {
-                        Object beanInstance = beanManager.getReference(configuration.getBean(), Object.class,
-                                beanManager.createCreationalContext(configuration.getBean()));
-
-                        if (configuration.getInvokerClass() != null) {
-                            try {
-                                Constructor<? extends Invoker> constructorUsingBeanInstance = configuration.getInvokerClass()
-                                        .getConstructor(Object.class);
-                                if (constructorUsingBeanInstance != null) {
-                                    mediator.setInvoker(constructorUsingBeanInstance.newInstance(beanInstance));
-                                } else {
-                                    mediator.setInvoker(configuration.getInvokerClass().getDeclaredConstructor()
-                                            .newInstance());
-                                }
 
-                            } catch (InstantiationException | IllegalAccessException e) {
-                                log.unableToCreateInvoker(configuration.getInvokerClass(), e);
-                                return;
+        // Before weaving hook
+        Executor executor = Runnable::run;
+        if (!hooks.isUnsatisfied()) {
+            Executor custom = hooks.get().before(registars, channelRegistry);
+            if (custom == null) {
+                System.out.println("Hook called by no executor");
```

Comment:
I see this error message in the logs, and similar errors to before which indicates that context propagation is still happening

Reply:
Thanks @kabir. So may do something wrong with CP to get a cleared executor...

Reply:
@cescoffier Ah! I debugged the CleanupContextPropagation stuff and I haven't set up the proper dependencies between the WildFly modules. I need to make the reactive messaging one aware of the CP one. I'll try that and get back to you

Reply:
Now I get the `custom` executor. It needs a small change as suggested in https://github.com/smallrye/smallrye-reactive-messaging/pull/716/files#r481279901.

I get another error:
```
2020-09-01 18:29:39,692 ERROR [io.smallrye.reactive.messaging.provider] (MSC service thread 1-3) SRMSG00212: Unable to initialize mediator: org.wildfly.test.integration.microprofile.reactive.messaging.sanity.Bean#toUpperCase: java.lang.IllegalStateException: No ReactiveStreamsFactory implementation found!
	at org.eclipse.microprofile.reactive-streams-operators.api@1.0.1//org.eclipse.microprofile.reactive.streams.operators.spi.ReactiveStreamsFactoryResolver.instance(ReactiveStreamsFactoryResolver.java:64)
	at org.eclipse.microprofile.reactive-streams-operators.api@1.0.1//org.eclipse.microprofile.reactive.streams.operators.ReactiveStreams.builder(ReactiveStreams.java:152)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.ProcessorMediator.processMethodReturningIndividualPayloadAndConsumingIndividualItem(ProcessorMediator.java:300)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.ProcessorMediator.initialize(ProcessorMediator.java:128)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.extension.MediatorManager.lambda$null$2(MediatorManager.java:214)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.extension.MediatorManager.lambda$initializeAndRun$3(MediatorManager.java:182)
	at io.smallrye.context-propagation@1.0.17-SNAPSHOT//io.smallrye.context.SmallRyeThreadContext.lambda$withContext$0(SmallRyeThreadContext.java:303)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.extension.MediatorManager.initializeAndRun(MediatorManager.java:180)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.extension.MediatorManager$Proxy$_$$_WeldClientProxy.initializeAndRun(Unknown Source)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.extension.ReactiveMessagingExtension.afterDeploymentValidation(ReactiveMessagingExtension.java:114)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:95)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.injection.MethodInvocationStrategy$SpecialParamPlusBeanManagerStrategy.invoke(MethodInvocationStrategy.java:187)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:330)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.event.ExtensionObserverMethodImpl.sendEvent(ExtensionObserverMethodImpl.java:123)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:308)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.event.ObserverMethodImpl.notify(ObserverMethodImpl.java:286)
	at javax.enterprise.api@2.0.2.redhat-00002//javax.enterprise.inject.spi.ObserverMethod.notify(ObserverMethod.java:124)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.util.Observers.notify(Observers.java:166)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.event.ObserverNotifier.notifySyncObservers(ObserverNotifier.java:285)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.event.ObserverNotifier.notify(ObserverNotifier.java:273)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.event.ObserverNotifier.fireEvent(ObserverNotifier.java:177)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.event.ObserverNotifier.fireEvent(ObserverNotifier.java:171)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.bootstrap.events.AbstractContainerEvent.fire(AbstractContainerEvent.java:53)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.bootstrap.events.AbstractDeploymentContainerEvent.fire(AbstractDeploymentContainerEvent.java:35)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.bootstrap.events.AfterDeploymentValidationImpl.fire(AfterDeploymentValidationImpl.java:28)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.bootstrap.WeldStartup.validateBeans(WeldStartup.java:505)
	at org.jboss.weld.core@3.1.4.Final-redhat-00001//org.jboss.weld.bootstrap.WeldBootstrap.validateBeans(WeldBootstrap.java:93)
	at org.jboss.as.weld@7.3.2.GA-redhat-SNAPSHOT//org.jboss.as.weld.WeldStartService.start(WeldStartService.java:98)
	at org.jboss.msc@1.4.11.Final-redhat-00001//org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1739)
	at org.jboss.msc@1.4.11.Final-redhat-00001//org.jboss.msc.service.ServiceControllerImpl$StartTask.execute(ServiceControllerImpl.java:1701)
	at org.jboss.msc@1.4.11.Final-redhat-00001//org.jboss.msc.service.ServiceControllerImpl$ControllerTask.run(ServiceControllerImpl.java:1559)
	at org.jboss.threads@2.3.3.Final-redhat-00001//org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads@2.3.3.Final-redhat-00001//org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1982)
	at org.jboss.threads@2.3.3.Final-redhat-00001//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
	at org.jboss.threads@2.3.3.Final-redhat-00001//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
	at java.base/java.lang.Thread.run(Thread.java:834)

2020-09-01 18:29:39,709 INFO  [stdout] (MSC service thread 1-3) ---> Creating output on Thread[MSC service thread 1-3,5,main]
2020-09-01 18:29:39,716 ERROR [io.smallrye.reactive.messaging.provider] (MSC service thread 1-3) SRMSG00200: The method org.wildfly.test.integration.microprofile.reactive.messaging.sanity.Bean#source has thrown an exception: io.smallrye.reactive.messaging.ProcessingException: SRMSG00103: Exception thrown when calling the method org.wildfly.test.integration.microprofile.reactive.messaging.sanity.Bean#source
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.AbstractMediator.lambda$initialize$0(AbstractMediator.java:68)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.AbstractMediator.invoke(AbstractMediator.java:79)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.PublisherMediator.produceAPublisherBuilderOfPayloads(PublisherMediator.java:110)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.PublisherMediator.initialize(PublisherMediator.java:70)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.extension.MediatorManager.lambda$null$2(MediatorManager.java:214)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.extension.MediatorManager.lambda$initializeAndRun$3(MediatorManager.java:182)
	at io.smallrye.context-propagation@1.0.17-SNAPSHOT//io.smallrye.context.SmallRyeThreadContext.lambda$withContext$0(SmallRyeThreadContext.java:303)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.extension.MediatorManager.initializeAndRun(MediatorManager.java:180)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.extension.MediatorManager$Proxy$_$$_WeldClientProxy.initializeAndRun(Unknown Source)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.extension.ReactiveMessagingExtension.afterDeploymentValidation(ReactiveMessagingExtension.java:114)
--- SNIP ---
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.smallrye.reactive.messaging@2.4.0-SNAPSHOT//io.smallrye.reactive.messaging.AbstractMediator.lambda$initialize$0(AbstractMediator.java:66)
	... 40 more
Caused by: java.lang.IllegalStateException: No ReactiveStreamsFactory implementation found!
	at org.eclipse.microprofile.reactive-streams-operators.api@1.0.1//org.eclipse.microprofile.reactive.streams.operators.spi.ReactiveStreamsFactoryResolver.instance(ReactiveStreamsFactoryResolver.java:64)
	at org.eclipse.microprofile.reactive-streams-operators.api@1.0.1//org.eclipse.microprofile.reactive.streams.operators.ReactiveStreams.of(ReactiveStreams.java:86)
	at deployment.rx-messaging.war//org.wildfly.test.integration.microprofile.reactive.messaging.sanity.Bean.source(Bean.java:43)
	at deployment.rx-messaging.war//org.wildfly.test.integration.microprofile.reactive.messaging.sanity.Bean$Proxy$_$$_WeldClientProxy.source(Unknown Source)
	... 45 more
```
That is probably more WildFly modules stuff, I will let you know (probably in the morning)

Reply:
It's because it does not find the Mutiny based implementation of Reactive Streams Operators

Reply:
This is odd, just before the error:
```
"MSC service thread 1-2"@2,744 in group "main": RUNNING
instance:57, ReactiveStreamsFactoryResolver {org.eclipse.microprofile.reactive.streams.operators.spi}
builder:152, ReactiveStreams {org.eclipse.microprofile.reactive.streams.operators}
processMethodReturningIndividualPayloadAndConsumingIndividualItem:300, ProcessorMediator {io.smallrye.reactive.messaging}
initialize:128, ProcessorMediator {io.smallrye.reactive.messaging}
lambda$null$2:214, MediatorManager {io.smallrye.reactive.messaging.extension}
accept:-1, 1860620473 {io.smallrye.reactive.messaging.extension.MediatorManager$$Lambda$834}
forEach:1541, ArrayList {java.util}
lambda$initializeAndRun$3:182, MediatorManager {io.smallrye.reactive.messaging.extension}
run:-1, 13993265 {io.smallrye.reactive.messaging.extension.MediatorManager$$Lambda$830}
lambda$withContext$0:303, SmallRyeThreadContext {io.smallrye.context}
execute:-1, 2104132501 {io.smallrye.context.SmallRyeThreadContext$$Lambda$829}
initializeAndRun:180, MediatorManager {io.smallrye.reactive.messaging.extension}
initializeAndRun:-1, MediatorManager$Proxy$_$$_WeldClientProxy {io.smallrye.reactive.messaging.extension}
afterDeploymentValidation:114, ReactiveMessagingExtension {io.smallrye.reactive.messaging.extension}
invoke0:-1, NativeMethodAccessorImpl {jdk.internal.reflect}
invoke:62, NativeMethodAccessorImpl {jdk.internal.reflect}
invoke:43, DelegatingMethodAccessorImpl {jdk.internal.reflect}
invoke:566, Method {java.lang.reflect}
invoke:95, StaticMethodInjectionPoint {org.jboss.weld.injection}
invoke:187, MethodInvocationStrategy$SpecialParamPlusBeanManagerStrategy {org.jboss.weld.injection}
sendEvent:330, ObserverMethodImpl {org.jboss.weld.event}
sendEvent:123, ExtensionObserverMethodImpl {org.jboss.weld.event}
sendEvent:308, ObserverMethodImpl {org.jboss.weld.event}
notify:286, ObserverMethodImpl {org.jboss.weld.event}
notify:124, ObserverMethod {javax.enterprise.inject.spi}
notify:166, Observers {org.jboss.weld.util}
notifySyncObservers:285, ObserverNotifier {org.jboss.weld.event}
notify:273, ObserverNotifier {org.jboss.weld.event}
fireEvent:177, ObserverNotifier {org.jboss.weld.event}
fireEvent:171, ObserverNotifier {org.jboss.weld.event}
fire:53, AbstractContainerEvent {org.jboss.weld.bootstrap.events}
fire:35, AbstractDeploymentContainerEvent {org.jboss.weld.bootstrap.events}
fire:28, AfterDeploymentValidationImpl {org.jboss.weld.bootstrap.events}
validateBeans:505, WeldStartup {org.jboss.weld.bootstrap}
validateBeans:93, WeldBootstrap {org.jboss.weld.bootstrap}
start:98, WeldStartService {org.jboss.as.weld}
startService:1739, ServiceControllerImpl$StartTask {org.jboss.msc.service}
execute:1701, ServiceControllerImpl$StartTask {org.jboss.msc.service}
run:1559, ServiceControllerImpl$ControllerTask {org.jboss.msc.service}
run:35, ContextClassLoaderSavingRunnable {org.jboss.threads}
safeRun:1982, EnhancedQueueExecutor {org.jboss.threads}
doRunTask:1486, EnhancedQueueExecutor$ThreadBody {org.jboss.threads}
run:1377, EnhancedQueueExecutor$ThreadBody {org.jboss.threads}
run:834, Thread {java.lang}
```
The TCCL is an instance of jdk.internal.loader.ClassLoaders.AppClassLoader. In WildFly this should be an instance of ModuleClassLoader, and I have never seen it be anything else :) I'll see if I can figure out what is wrong.